src/parser/jobs/dnc/modules/DirtyDancing.tsx

```diff
@@ -0,0 +1,298 @@
+import {t} from '@lingui/macro'
+import {Plural, Trans} from '@lingui/react'
+import _ from 'lodash'
+import React, {Fragment} from 'react'
+import {Icon, Message} from 'semantic-ui-react'
+
+import {ActionLink, StatusLink} from 'components/ui/DbLink'
+import {RotationTable} from 'components/ui/RotationTable'
+import ACTIONS from 'data/ACTIONS'
+import STATUSES from 'data/STATUSES'
+import {CastEvent, DamageEvent} from 'fflogs'
+import Module, {dependency} from 'parser/core/Module'
+import CheckList, {Requirement, Rule} from 'parser/core/modules/Checklist'
+import Combatants from 'parser/core/modules/Combatants'
+import Invulnerability from 'parser/core/modules/Invulnerability'
+import Suggestions, {SEVERITY, TieredSuggestion} from 'parser/core/modules/Suggestions'
+import Timeline from 'parser/core/modules/Timeline'
+
+const ISSUE_SEVERITY_TIERS = {
+	1: SEVERITY.MINOR,
+	3: SEVERITY.MEDIUM,
+	5: SEVERITY.MAJOR,
+}
+
+const STEPS = [
+	ACTIONS.STANDARD_STEP.id,
+	ACTIONS.TECHNICAL_STEP.id,
+]
+
+const DANCE_MOVES = [
+	ACTIONS.ENTRECHAT.id,
+	ACTIONS.EMBOITE.id,
+	ACTIONS.JETE.id,
+	ACTIONS.PIROUETTE.id,
+]
+
+const STANDARD_FINISHES = [
+	ACTIONS.STANDARD_FINISH.id,
+	ACTIONS.SINGLE_STANDARD_FINISH.id,
+	ACTIONS.DOUBLE_STANDARD_FINISH.id,
+]
+
+const TECHNICAL_FINISHES = [
+	ACTIONS.TECHNICAL_FINISH.id,
+	ACTIONS.SINGLE_TECHNICAL_FINISH.id,
+	ACTIONS.DOUBLE_TECHNICAL_FINISH.id,
+	ACTIONS.TRIPLE_TECHNICAL_FINISH.id,
+	ACTIONS.QUADRUPLE_TECHNICAL_FINISH.id,
+]
+
+const FINISHES = [
+	...STANDARD_FINISHES,
+	...TECHNICAL_FINISHES,
+]
+
+const EXPECTED_DANCE_MOVE_COUNT = {
+	[ACTIONS.DOUBLE_STANDARD_FINISH.id]: 2,
+	[ACTIONS.QUADRUPLE_TECHNICAL_FINISH.id]: 4,
+}
+
+class Dance {
+	start: number
+	end?: number
+	rotation: CastEvent[] = []
+	dancing: boolean = false
+	resolved: boolean = false
+
+	dirty: boolean = false
+	missed: boolean = false
+	footloose: boolean = false
+
+	public get error(): boolean {
+		return this.dirty || this.missed || this.footloose
+	}
+
+	public get expectedFinishId(): number {
+		const actualFinish = _.last(this.rotation)
```

Comment:
would it make more sense to base the expected finish on the start move?

Reply:
The problem with that is the correct opener begins with a Double Standard Finish that was set up pre-pull, so we can't go off the starter in that case. Given that, I figured it'd be easier to just always infer the proper finish from the one we actually got.

Reply:
Good point; hadn't thought of that. Makes sense this way around then.